hal/mbed_sleep_manager.c

```diff
@@ -14,18 +14,139 @@
 #include <limits.h>
+#include <stdio.h>
 
 #if DEVICE_SLEEP
 
 // deep sleep locking counter. A target is allowed to deep sleep if counter == 0
 static uint16_t deep_sleep_lock = 0U;
 
-void sleep_manager_lock_deep_sleep(void)
+#ifdef MBED_SLEEP_TRACING_ENABLED
+
+// Length of the identifier extracted from the driver name to store for logging.
+#define IDENTIFIER_WIDTH 7
+// Number of drivers that can be stored in the structure
+#define STATISTIC_COUNT  10
+
+typedef struct sleep_statistic {
+    char identifier[IDENTIFIER_WIDTH];
+    uint8_t count;
+} sleep_statistic_t;
+
+static sleep_statistic_t sleep_stats[STATISTIC_COUNT];
+
+static const char* strip_path(const char* const filename)
+{
+    char *output = strrchr(filename, '/');
+
+    if (output != NULL) {
+        return output + 1;
+    }
+
+    output = strrchr(filename, '\\');
+
+    if (output != NULL) {
+        return output + 1;
+    }
+
+    return filename;
+}
+
+static sleep_statistic_t* sleep_tracker_find(const char *const filename)
+{
+    char temp[IDENTIFIER_WIDTH];
+    strncpy(temp, filename, IDENTIFIER_WIDTH);
+    temp[IDENTIFIER_WIDTH - 1] = '\0';
+
+    // Search for the a driver matching the current name and return it's index
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (strcmp(sleep_stats[i].identifier, temp) == 0) {
+            return &sleep_stats[i];
+        }
+    }
+
+    return NULL;
+}
+
+static sleep_statistic_t* sleep_tracker_add(const char* const filename)
+{
+    char temp[IDENTIFIER_WIDTH];
+    strncpy(temp, filename, IDENTIFIER_WIDTH);
+    temp[IDENTIFIER_WIDTH - 1] = '\0';
+
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (sleep_stats[i].identifier[0] == '\0') {
+            core_util_critical_section_enter();
+            strncpy(sleep_stats[i].identifier, temp, sizeof(temp));
+            core_util_critical_section_exit();
+
+            return &sleep_stats[i];
+        }
+    }
+
+    // Panic if there are no free indexes left to track with
+    error("No free indexes left to use in mbed stats tracker");
+
+    return NULL;
+}
+
+static void sleep_tracker_print_stats(void)
+{
+    for (int i = 0; i < STATISTIC_COUNT; ++i) {
+        if (sleep_stats[i].count == 0) {
+            continue;
+        }
+
+        if (sleep_stats[i].identifier[0] == '\0') {
+            return;
+        }
+
+        printf("[id: %s, count: %u]\r\n", sleep_stats[i].identifier,
+                                          sleep_stats[i].count);
+    }
+}
+
+void sleep_tracker_lock(const char* const filename, int line)
+{
+    const char* const stripped_path = strip_path(filename);
+
+    sleep_statistic_t* stat = sleep_tracker_find(stripped_path);
+
+    // Entry for this driver does not exist, create one.
+    if (stat == NULL) {
+        stat = sleep_tracker_add(filename);
+    }
+
+    core_util_atomic_incr_u8(&stat->count, 1);
+
+    printf("LOCK: %s, ln: %i, lock count: %u\r\n", stripped_path, line, deep_sleep_lock);
```

Comment:
Why is this printf and not debug? If debug it could be stripped from a release build.